components/raftstore-v2/src/batch/store.rs

```diff
@@ -662,45 +663,71 @@ impl<EK: KvEngine, ER: RaftEngine> StoreSystem<EK, ER> {
             let logger = self.logger.clone();
             let router = router.clone();
             let registry = tablet_registry.clone();
-            let limiter = Limiter::new(MIN_MANUAL_FLUSH_RATE);
-            let mut max_rate = cfg.value().max_manual_flush_rate;
-            if max_rate < MIN_MANUAL_FLUSH_RATE {
-                max_rate = MIN_MANUAL_FLUSH_RATE;
-            }
-            worker.spawn_interval_task(cfg.value().raft_engine_purge_interval.0, move || {
-                let _guard = WithIoType::new(IoType::RewriteLog);
-                match raft_clone.manual_purge() {
-                    Ok(mut regions) => {
-                        if regions.is_empty() {
-                            return;
+            let base_max_rate = cfg
+                .value()
+                .max_manual_flush_rate
+                .clamp(MIN_MANUAL_FLUSH_RATE, f64::INFINITY);
+            let mut last_flush = (
+                EK::get_accumulated_flush_count().unwrap(),
+                TiInstant::now_coarse(),
+            );
+            worker.spawn_interval_async_task(cfg.value().raft_engine_purge_interval.0, move || {
+                let regions = {
+                    let _guard = WithIoType::new(IoType::RewriteLog);
+                    match raft_clone.manual_purge() {
+                        Err(e) => {
+                            warn!(logger, "purge expired files"; "err" => %e);
+                            Vec::new()
+                        }
+                        Ok(regions) => regions,
+                    }
+                };
+                // Lift up max rate if the background flush rate is high.
+                let flush_count = EK::get_accumulated_flush_count().unwrap();
+                let now = TiInstant::now_coarse();
+                let duration = now.saturating_duration_since(last_flush.1).as_secs_f64();
+                let max_rate = if duration > 10.0 {
+                    let total_flush_rate = (flush_count - last_flush.0) as f64 / duration;
+                    last_flush = (flush_count, now);
+                    base_max_rate.clamp(total_flush_rate, f64::INFINITY)
```

Comment:
Seems base_max_rate can only be increased over time. Is it intentional?

Reply:
Yes, refer to PR description.

Reply:
If flush is very high just at some time, the calculated high base_max_rate may be never changed and will be used for all following region picks even when flush is low.

Reply:
Why is that? `max_rate` is recalculated every interval.

Reply:
ok, I misunderstanded clamp.